### DISCUSSION TITLE
expressionScope传入的属性用到外部state，当 state 改变，但是expressionScope里边传入的属性输出的还是旧state

### DISCUSSION THREAD

**Discussion by XionGuo37** [alibaba/formily#1567] [Q&A 1.x] [answered]
https://codesandbox.io/s/gallant-faraday-22879?file=/index.js:1436-1451

**Comment by liuweiGL** [comment, marked as answer]
https://codesandbox.io/s/musing-taussig-oyun2

**Comment by liuweiGL** [comment]
这属于 react 的基础知识，原因是你的 handleClick 访问到的是闭包中的 msg。
解决办法：
1. 把 handleXXX 使用 useCallback 包一下，expressionScope 使用 useMemo 包装。
2. 使用 ref 代替 state

**Reply by XionGuo37** [reply]
瑞思拜
